File: 1448.java
class Solution {
    private int numNodes = 0;
    private void findNumGoodNodes(TreeNode root, int maxTillNow) {
        if (root == null) {
            return;
        }
        if (root.val >= maxTillNow) {
            this.numNodes += 1;
        }
        maxTillNow = Math.max(maxTillNow, root.val);
        findNumGoodNodes(root.left, maxTillNow);
        findNumGoodNodes(root.right, maxTillNow);
    } 
    public int goodNodes(TreeNode root) {
        findNumGoodNodes(root, -100005);
        return this.numNodes;
    }
}

File: 1448.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findGoodNodes(TreeNode* root, int largest) {
        if (root == nullptr) {
            return 0;
        }
        
        int l, r; 
        l = findGoodNodes(root->left, max(root->val, largest));
        r = findGoodNodes(root->right, max(root->val, largest));
        if (root->val >= largest) {
            return l + r + 1;
        }
        return l + r;
    }
    int goodNodes(TreeNode* root) {
        return findGoodNodes(root, root->val);
    }
};
